# **`Simple Neural Network` in Python using TensorFlow**

#### **1. Importing the required libraries.**

In [1]:
import pandas as pd
import seaborn as sns
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import tensorflow as tf
# import keras

## **2. Data Preprocessing:**
(Steps before Creating a Neural Network)

In [2]:
#remove warnings
import warnings
warnings.filterwarnings('ignore')

#### **2.1 Load the Titanic Dataset**

In [3]:
# Load Titanic dataset
titanic = sns.load_dataset('titanic')

#### **2.2 Dropping Missing Values**

In [4]:
# Preprocessing
# Dropping rows with missing 'age' and 'embarked' values
titanic.dropna(subset=['age', 'embarked'], inplace=True)

#### **2.3 Convert Categorical Data to Numerical Data**

In [5]:
# Converting categorical variables to dummy variables
titanic = pd.get_dummies(titanic, columns=['sex', 'embarked', 'class', 'who', 'deck'], drop_first=True)

#### **2.4 Splitting the Data into Features (X) and Targets (y)**

In [6]:
# Selecting features and target
X = titanic.drop(['survived', 'alive', 'embark_town', 'adult_male', 'alone'], axis=1) # axis = 1 means drop columns & axis = 0 means drop rows
y = titanic['survived']

#### **2.5 Splitting the Data into Training and Testing Sets**

In [7]:
# Splitting the dataset
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42) 
# 80% training and 20% test, random_state fixed the randomness of the split

#### **2.6 Standardizing the Data**

In [8]:
# Standardizing the data
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

In [9]:
X_train.shape[1]

18

Now at this stage we have preprocessed the data and now we can move to the next step of creating a Neural Network.

## **3. Building the Model**

#### **3.1 Define the Layers of the Model**

In [10]:
# Building the Simple Neural Network (Model)

# define the layers of the model
input_layer = tf.keras.layers.Dense(10, activation='relu', input_shape=(X_train.shape[1],)) # input layer, 10 neurons, relu activation function, input shape = number of features
# > Dense layer is a fully connected layer

# hidden_layer = tf.keras.layers.Dense(10, activation='relu') # hidden layer, 10 neurons, relu activation function
output_layer = tf.keras.layers.Dense(1, activation='sigmoid') # output layer, 1 neuron, sigmoid activation function

This code is defining the architecture of a simple neural network using TensorFlow's Keras API. Here's a breakdown of what's happening:

1. **Input Layer**: The first line of code is defining the input layer of the neural network. This layer will have 10 neurons (also known as nodes), and it uses the ReLU (Rectified Linear Unit) activation function. The `input_shape` parameter is set to the number of features in the training data (`X_train.shape[1]`), which tells the network how many inputs to expect.

    ```python
    input_layer = tf.keras.layers.Dense(10, activation='relu', input_shape=(X_train.shape[1],))
    ```

- The `Dense` function is used to create a fully connected layer, meaning each neuron in this layer will be connected to all neurons in the next layer. 
- The `ReLU` activation function is commonly used in neural networks to introduce non-linearity into the model. It outputs the input directly if it's positive; otherwise, it outputs zero.

2. **Output Layer**: The last line of code is defining the output layer of the neural network. This layer has just `1 neuron`, and it uses the `sigmoid` activation function.

    ```python
    output_layer = tf.keras.layers.Dense(1, activation='sigmoid')
    ```

- The `sigmoid activation` function is commonly used for binary classification problems. It outputs a value between 0 and 1, which can be interpreted as the probability of the positive class in a binary classification problem.

The hidden layer code is commented out in this snippet, but if it were included, it would add an additional layer of 10 neurons using the ReLU activation function between the input and output layers.

> In summary, this code is setting up a simple neural network with one input layer and one output layer. The input layer takes in the features from the training data and passes them through the network. The output layer then makes a prediction based on these features.

#### **3.2 Combine the Layers to Create the Model**

In [11]:
# combine the layers into a model
model = tf.keras.models.Sequential([input_layer, 
                                    # hidden_layer, 
                                    output_layer])

This piece of code is creating a `Sequential model` in TensorFlow's Keras API. A Sequential model is a linear stack of layers, where you can just add one layer after another.

Here's a breakdown of what's happening:

1. `tf.keras.models.Sequential`: This initializes a new Sequential model. This is the simplest kind of Keras model for neural networks that are just composed of a single stack of layers connected sequentially.

2. `[input_layer, output_layer]`: This is a list of the layers that you want to add to the model. The order in which you add them is the order in which the data will flow through the network. In this case, the data will first go through the `input_layer`, and then it will go to the `output_layer`.

The `hidden_layer` is commented out in this code, but if it were included, it would add an additional layer between the input and output layers.

> In summary, this code is creating a new Sequential model and adding the input and output layers to it. The resulting `model` object represents your neural network, and it can be trained on your data using its `fit` method.

#### **3.3 Compile the Model**

In [12]:
# Compile the model
model.compile(optimizer='adam', 
              loss='binary_crossentropy', 
              metrics=['accuracy'])

This piece of code is compiling the model. Compiling the model means `configuring the learning process before training the model`. It requires an optimizer, a loss function, and optionally, some metrics to monitor during training and testing.

Here's a breakdown of what's happening:

1. **optimizer='`adam`'**: The optimizer controls the learning rate. In this case, '`adam`' is used, which is an algorithm for `first-order gradient-based optimization of stochastic objective functions`. It's a popular choice because it works well in practice and requires little configuration.

2. **loss='`binary_crossentropy`'**: This is the `loss function that the model will try to minimize`. For binary classification problems like this one, '`binary_crossentropy`' is a common choice. It's a measure of error for binary classification problems.

3. **metrics=['accuracy']**`: This is a list of metrics to be evaluated by the model during training and testing. In this case, we're interested in the 'accuracy' of the model, which is the proportion of correct predictions.

> In summary, this code is setting up the learning process by specifying the optimizer, loss function, and metrics for the model. After this step, the model will be ready to be trained with data.

## **4. Training the Model**

In [13]:
%%time
# Training the model
model.fit(X_train, y_train, epochs=100, batch_size=32, verbose=1)

Epoch 1/100


18/18 [==============================] - 4s 7ms/step - loss: 0.8636 - accuracy: 0.3849
Epoch 2/100
18/18 [==============================] - 0s 6ms/step - loss: 0.7951 - accuracy: 0.4622
Epoch 3/100
18/18 [==============================] - 0s 5ms/step - loss: 0.7410 - accuracy: 0.5185
Epoch 4/100
18/18 [==============================] - 0s 5ms/step - loss: 0.6945 - accuracy: 0.6766
Epoch 5/100
18/18 [==============================] - 0s 5ms/step - loss: 0.6564 - accuracy: 0.7083
Epoch 6/100
18/18 [==============================] - 0s 6ms/step - loss: 0.6235 - accuracy: 0.7258
Epoch 7/100
18/18 [==============================] - 0s 5ms/step - loss: 0.5949 - accuracy: 0.7487
Epoch 8/100
18/18 [==============================] - 0s 5ms/step - loss: 0.5694 - accuracy: 0.7540
Epoch 9/100
18/18 [==============================] - 0s 6ms/step - loss: 0.5463 - accuracy: 0.7680
Epoch 10/100
18/18 [==============================] - 0s 6ms/step - loss: 0.5258 - accuracy: 0.7856
Epoch 11/100
18/18 [=

This code is training the model using the `fit` method. Here's a breakdown of what's happening:

1. **model.fit(X_train, y_train, epochs=100, batch_size=32, verbose=1)**: 
   - This line is training the model for 100 epochs on the training data (`X_train` and `y_train`). 
     - `epochs`: An epoch is one complete pass through the training data. The model will train on the entire training data for 100 epochs, which means it will see each sample in the training data 100 times.
   - The `batch_size` parameter is set to 32, which means that the model will update its weights after every 32 samples. The `verbose` parameter is set to 1, which means that the method will output progress logs after each epoch.
   - The `verbose=1` parameter controls the amount of information that's printed during training. Setting it to 1 means that the method will output progress logs after each epoch.

**Observations from the output:**
- The output shows the progress of the training process. 
- For each epoch, it shows the loss and accuracy of the model on the training data. 
- The loss is a measure of the model's error, and the accuracy is the proportion of correct predictions. 
- The goal of the training process is to minimize the loss and maximize the accuracy.

> In summary, this code is training the model on the training data for 100 epochs, updating the model's weights after every 32 samples, and outputting progress logs after each epoch. The output shows that the model's accuracy on the training data is improving over time, which is a good sign that the model is learning from the data.

## **5. Evaluating the Model**

In [14]:
# Evaluating the model
loss, accuracy = model.evaluate(X_test, y_test, verbose=1)
print(f"Test Accuracy: {accuracy}")
print(f"Test Loss: {loss}")

5/5 [==============================] - 1s 8ms/step - loss: 0.5159 - accuracy: 0.7483
Test Accuracy: 0.748251736164093
Test Loss: 0.5159400701522827


This piece of code is evaluating the performance of the trained model on the test data (`X_test` and `y_test`). Here's a breakdown of what's happening:

1. `model.evaluate(X_test, y_test, verbose=1)`: This line is evaluating the model's performance on the test data. The `evaluate` method returns the loss value and metrics values for the model in test mode. In this case, the method returns the loss and accuracy because these were specified when compiling the model.

2. `loss, accuracy = model.evaluate(X_test, y_test, verbose=1)`: This line is unpacking the returned values into the variables `loss` and `accuracy`.

3. `print(f"Test Accuracy: {accuracy}")` and `print(f"Test Loss: {loss}")`: These lines are printing the test accuracy and test loss.

**Observations from the output:**
- The output shows the loss and accuracy of the model on the test data. 
- The loss is a measure of the model's error, and the accuracy is the proportion of correct predictions. 
- In this case, the model has an accuracy of about `74.8%` and a loss of about `0.516` on the test data.
- Which means the model is able to predict the correct class about 74.8% of the time on the test data.

> In summary, this code is evaluating the model's performance on the test data and printing the results. The output shows that the model's accuracy on the test data is slightly lower than its accuracy on the training data, which is a common occurrence and can indicate that the model is overfitting to the training data.

## **Let's see all the steps in action within one snippet of code:**

In [15]:
%%time
import tensorflow as tf
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import pandas as pd
import seaborn as sns

#1 remove warnings
import warnings
warnings.filterwarnings('ignore')

#2 Load Titanic dataset
titanic = sns.load_dataset('titanic')

#3 Preprocessing
#3.1 Dropping rows with missing 'age' and 'embarked' values
titanic.dropna(subset=['age', 'embarked'], inplace=True)

#3.2 Converting categorical variables to dummy variables
titanic = pd.get_dummies(titanic, columns=['sex', 'embarked', 'class', 'who', 'deck'], drop_first=True)

#4 Selecting features and target
X = titanic.drop(['survived', 'alive', 'embark_town', 'adult_male', 'alone'], axis=1)
y = titanic['survived']

#5 Splitting the dataset
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

#6 Standardizing the data
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

#7 Building the model
input_layer = tf.keras.layers.Dense(10, activation='relu', input_shape=(X_train.shape[1],)) # input layer
# hidden_layer = tf.keras.layers.Dense(10, activation='relu') # hidden layer
output_layer = tf.keras.layers.Dense(1, activation='sigmoid') # output layer

model = tf.keras.models.Sequential([input_layer, 
                                    # hidden_layer, 
                                    output_layer])

#8 Compile the model
model.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])

#9 Training the model
model.fit(X_train, y_train, epochs=100, batch_size=32, verbose=1)

#10 Evaluating the model
loss, accuracy = model.evaluate(X_test, y_test, verbose=1)
print(f"Test Accuracy: {accuracy}")
print(f"Test Loss: {loss}")

Epoch 1/100


18/18 [==============================] - 4s 6ms/step - loss: 0.7363 - accuracy: 0.5940
Epoch 2/100
18/18 [==============================] - 0s 6ms/step - loss: 0.6832 - accuracy: 0.6450
Epoch 3/100
18/18 [==============================] - 0s 8ms/step - loss: 0.6417 - accuracy: 0.6837
Epoch 4/100
18/18 [==============================] - 0s 5ms/step - loss: 0.6074 - accuracy: 0.7030
Epoch 5/100
18/18 [==============================] - 0s 5ms/step - loss: 0.5792 - accuracy: 0.7188
Epoch 6/100
18/18 [==============================] - 0s 6ms/step - loss: 0.5538 - accuracy: 0.7417
Epoch 7/100
18/18 [==============================] - 0s 5ms/step - loss: 0.5321 - accuracy: 0.7663
Epoch 8/100
18/18 [==============================] - 0s 5ms/step - loss: 0.5133 - accuracy: 0.7680
Epoch 9/100
18/18 [==============================] - 0s 11ms/step - loss: 0.4969 - accuracy: 0.7715
Epoch 10/100
18/18 [==============================] - 0s 8ms/step - loss: 0.4828 - accuracy: 0.7786
Epoch 11/100
18/18 [

# **Assignment:** Plot the Training and Validation Accuracy and Loss for each epoch

## **Check if `GPU` is available**

In [17]:
import sys

import tensorflow.keras
import pandas as pd
import sklearn as sk
import scipy as sp
import tensorflow as tf
import platform

print(f"Python Platform: {platform.platform()}")
print(f"Tensor Flow Version: {tf.__version__}")
print()
print(f"Python {sys.version}")
print(f"Pandas {pd.__version__}")
print(f"Scikit-Learn {sk.__version__}")
print(f"SciPy {sp.__version__}")
gpu = len(tf.config.list_physical_devices('GPU'))>0
print("GPU is", "available" if gpu else "NOT AVAILABLE")

Python Platform: Windows-10-10.0.19045-SP0
Tensor Flow Version: 2.10.0

Python 3.8.15 | packaged by conda-forge | (default, Nov 22 2022, 08:43:00) [MSC v.1929 64 bit (AMD64)]
Pandas 2.0.3
Scikit-Learn 1.3.0
SciPy 1.10.1
GPU is NOT AVAILABLE


---